src/evict.c

```diff
@@ -726,8 +729,18 @@ int performEvictions(void) {
         }
     }
     /* at this point, the memory is OK, or we have reached the time limit */
+    if (server.maxmemory_reserved_scale && server.maxmemory_available) {
+        size_t mem_used = zmalloc_used_memory();
+        size_t overhead = freeMemoryGetNotCountedMemory();
+        mem_used = (mem_used > overhead) ? mem_used - overhead : 0;
+        if (mem_used > server.maxmemory) {
+            result = EVICT_FAIL;
+            goto cant_free;
+        }
+    }
```

Comment:
I don't think we should check `server.maxmemory_reserved_scale` and `server.maxmemory_available`. Also we should not `goto cant_free` on L725/728; instead, we should just break out of the while loop there and then always compute the `result` based on the current memory state. Otherwise, failing to evict keys due to used memory being greater than the new "soft" limit will always result in EVICT_FAIL/OOM rejection.

```suggestion
    size_t mem_used = zmalloc_used_memory();
    size_t overhead = freeMemoryGetNotCountedMemory();
    mem_used = (mem_used > overhead) ? mem_used - overhead : 0;
    if (mem_used > server.maxmemory) {
        result = EVICT_FAIL;
        goto cant_free;
    }
```

Reply:
Hi Ping, I think you misunderstand the logic for this condition: server.maxmemory_reserved_scale && server.maxmemory_available. let me explain to you here:
There are 2 cases:
Case 1. server.maxmemory_available == server.maxmemory.  
In this case, if getMaxmemoryState() return C_ERR, server should make the key eviction. then in the following for... loop, some keys are evicted and some memory are released. Basiclly, we can guarantee used_memory is below soft max.  We do not need to calculate the used_memory again and just keep current logic.

Case 2.  server.maxmemory_available < server.maxmemory. 
In this case, even we make some key eviction behavior, but we are not sure if used_memory is smaller than the soft memory, then we should calculate here.

Thus, only when server.maxmemory_reserved_scale is set, we need to recalculate the used_memory （I think keep one conidtion if (server.maxmemory_reserved_scale) is enough because server.maxmemory_available is always true）